keps/sig-cluster-lifecycle/kubeadm/20190925-component-configs.md

```diff
@@ -0,0 +1,223 @@
+- Providing support for validation by vendoring component specific code can lead to false-positive and missed errors. This can be caused by the deviation between the imported code by kubeadm and the actual component.
+- Migrating between config versions also requires code to be vendored in or forked from the components. This can also lead to difficult to diagnose errors and changed behaviour in unexpected ways.
+- Not distinguishing between kubeadm generated and user supplied component configs, necessitates the requirement for config migration. Kubeadm generated configs that are not modified by users directly in the config map, can be fully regenerated upon config format change. In all other cases kubeadm must convert the config to preserve user changes.
+
+These observations along with the emergence of kubeadm phases, kustomize support and work around addon operators allows us to have the following future goals.
+
+### Goals
+
+- Stop vendoring internal types and validation, defaulting and conversion code.
+- Stop defaulting component configs.
+- Outsource validation to the component binaries themselves.
+- Distinguish between user supplied and kubeadm generated and unmodified component configs.
+- Eradicate the need to convert configs
+- Use more community wide solutions, like Kustomize, kubeadm phases and addon operators
+
+### Non-Goals
+
+- Change the handling of kubeadm’s own config types in any way.
+- Cover in any way the migration of command line flags to component configs.
+- Define strategies for config map naming and backup.
+
+## Proposal
+
+This is where we get down to the nitty gritty of what the proposal actually is.
+
+### Stop Component Config defaulting
+
+Component config defaulting has a number of issues:
+- It’s difficult to tell in retrospect if a setting value was intended or simply defaulted.
+- If a default value changes a defaulted setting will stop us from using the new value.
+- Defaulting bloats the resulting config, thus making it hard for users to migrate it upon a version change.
+
+Hence, kubeadm should stop defaulting component configs and leave this to components themselves.
+
+### Delegate config validation to the components
+
+Currently, kubeadm vendors in pieces of Go code from the components to be able to perform component config validation. This is a bad practice as it bloats the kubeadm dependencies and deviation between the vendored code and the components can lead to inconsistencies.  
+To solve this, validation is going to be performed by executing a container or the binary of the component to be used with an appropriate command to verify the supplied config.  
+If a target component does not have any means to verify its config, then kubeadm won’t do any validation for the time being. It will instead display a warning to the user denoting that certain component’s config is unvalidated.
```

Comment:
given kubeadm supports multiple versions of components, we need to track the minimum supported version of a component that added e.g. a `--validate` flag.